chapter-6/exercises.pl

```diff
@@ -100,37 +100,13 @@
 %% identical to List2, except that the first and last elements are
 %% exchanged. Hint: here's where append comes in useful again.
 
-%% Helper function
-
-%% equal
-%% base case
-equal([],[]).
-%% inductive case
-equal([H1|T1], [H1|T2]) :- equal(T1,T2).
-
-%% reverse
-%% base case
-accRev([],A,A).
-%% inductive case
-accRev([H|T],A,R) :- accRev(T,[H|A],R).
-%% main
-rev(L,R) :- accRev(L,[],R).
-
-chopOffEnds([_|T],Outlist) :-
-  rev(T, [_|RevOutlist]),
-  rev(RevOutlist,Outlist).
-
-%% Actual function
-
-%% swapfl(List1, List2)
-swapfl(List1,List2) :-
-  compareFirstLast(List1,List2),
-  compareFirstLast(List2,List1),
-  chopOffEnds(List1, Outlist1),
-  chopOffEnds(List2, Outlist2),
-  equal(Outlist1,Outlist2).
-
-compareFirstLast([H|T1],L2) :- rev(L2, [H|_]).
+%% As per strict interpretation of the question. Empty lists or lists with one
+%% element cannot be compared since they are missing the first and last element.
```

Comment:
As per my interpretation of lists in Prolog and other languages a list with one element is basically the element cons with empty list. That is [a] would a :: [] or something like cons(a, []). So in this case, the first element of a list with single element would always be the head element and last element or tail element would be the empty list []. A list with single element like [a] as per my understanding cannot be represented as [] :: a.
In short if you want to divide a list with single element in two parts, then you can do so. Also those two parts cannot be interchanged as per definition. That is an empty list cannot be first element in a list, unless it is a list of list.

However you are right. If the task description mentions that lists with single element are supposed to be taken into account, then the above solution would fall apart. The solution is not generic enough.